### DISCUSSION TITLE
Is it possible to output 16-bit uncompressed TIFF?

### DISCUSSION THREAD

**Discussion by mstenta** [OpenStitching/stitching#224] [Q&A]
Thank you for this library! It works like magic and I'm so grateful for the work you've done. :-)

I have one question: is it possible to configure or disable compression in the final result image file?

I am experimenting with using this library to stitch together four high quality uncompressed TIFF image files.

Each is 2602x3908 pixels and 30.6 megabytes. There is a lot of overlap between them, with a combined result size of 4729x4628 pixels. Nevertheless, the combined TIFF file size is only 26.4 megabytes. When I look at the EXIF data I see that LZW compression is applied.

```
$ exiftool -compression result.tif 
Compression                     : LZW
```

For JPEGs there seems to be a much larger loss of information. The same four images in JPEG format are 10.1 megabytes each, but the combined result JPEG is 4.7 megabytes.

Are there any options for configuring or disabling this so that the resulting image is uncompressed? I don't see anything in the `-h` help options, but maybe I'm missing it.

If not, would this be a good candidate for a feature request?

**Comment by lukasalexanderweber** [comment]
The short answer is that the output image name is directly passed into OpenCVs `imwrite`, so there is no logic for that within `stitching`

**Comment by mstenta** [comment]
Related but separate question (perhaps deserving of it's own discussion): I noticed that 16-bit TIFFs always end up as 8-bit TIFFs in the result. Would it be possible to support 16-bit TIFF output?

A little more background, to explain why I'm interested in these things: I am experimenting with digitizing large format (4x4 inch) film negatives using a DSLR camera over a light table, with the end goal of producing very large uncompressed TIFF files that can be professionally printed 30x30 inches or larger. I take multiple RAW photographs of the negative in a grid pattern, then use [Darktable](https://www.darktable.org/) to invert, color correct, etc, and export 16-bit TIFF images. My hope is to stitch these TIFFs together into a final high-quality file that can be used as the "final" image, which can be used to produce derivatives for print/web/etc.

**Reply by lukasalexanderweber** [reply]
In the [original study](https://henry.baw.de/handle/20.500.11970/112696) we worked with binary tiffs. We did some postprocessing to generate the binary tif result. As far as I remember we stored the stitching result in an uncompressed Format (PNG?) and used another Software for compression

**Comment by mstenta** [comment]
> The short answer is that the output image name is directly passed into OpenCVs `imwrite`, so there is no logic for that within `stitching`

I see. So this is an upstream limitation of OpenCV's `imwrite`? Does that provide any options? (I will look into it myself too.)

Thank you for the quick answer!

**Comment by mstenta** [comment]
Researching `imwrite()`...

From https://docs.opencv.org/4.x/d4/da8/group__imgcodecs.html (only including TIFF bullet point):

> Saves an image to a specified file.
> 
> The function imwrite saves the image to the specified file. The image format is chosen based on the filename extension (see [cv::imread](https://docs.opencv.org/4.x/d4/da8/group__imgcodecs.html#gacbaa02cffc4ec2422dfa2e24412a99e2) for the list of extensions). In general, only 8-bit unsigned (CV_8U) single-channel or 3-channel (with 'BGR' channel order) images can be saved using this function, with these exceptions:
> 
> - ...
> - With TIFF encoder, 8-bit unsigned (CV_8U), 16-bit unsigned (CV_16U), 32-bit float (CV_32F) and 64-bit float (CV_64F) images can be saved.
>   - Multiple images (vector of [Mat](https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html)) can be saved in TIFF format (see the code sample below).
>   - 32-bit float 3-channel (CV_32FC3) TIFF images will be saved using the LogLuv high dynamic range encoding (4 bytes per pixel)
> 
> If the image format is not supported, the image will be converted to 8-bit unsigned (CV_8U) and saved that way.

There is also mention of customizing compression in an example:

> The sample below shows how to create a BGRA image, how to set custom compression parameters and save it to a PNG file. It also demonstrates how to save multiple images in a TIFF file: 

Perhaps `stitching` could accept some CLI arguments that can be passed through to `imwrite()` to accomplish 16-bit uncompressed TIFF output? :thinking:

**Reply by lukasalexanderweber** [reply]
If this could elegantly be solved with one Parameter for all file formats (`kwargs` for `imwrite` as e.g. `output_parameters`?) this could be an interesting feature

**Comment by mstenta** [comment]
Documenting more research links...

- https://itecnotes.com/tecnote/c-writing-16-bit-uncompressed-image-using-opencv/
- https://answers.opencv.org/question/56594/how-to-use-imwrite-with-no-compression-to-save-tiff-images/

Also, according to [this SO answer](https://stackoverflow.com/a/18461475):

> By default `cv2.imread` will convert a 16 bit, three channel image in `a.tif` to 8 bit as shown in the question.

However, it looks like `imread()` has different options in OpenCV 4.x (used by `stitching`), so maybe this is no longer the case. It is worth confirming, otherwise there is a risk that data would be lost even if a 16-bit output is produced (16bit converted to 8bit internally, then converted back to 16bit).

**Comment by mstenta** [comment]
(I updated the discussion title to make this specific to "16-bit uncompressed TIFF output". But if a feature is added ideally it would support more options as well.)

**Comment by lukasalexanderweber** [comment]
From chatgpt:

```
import cv2
import argparse

# Initialize the ArgumentParser
parser = argparse.ArgumentParser(description='Save an image with specified compression parameters.')
parser.add_argument('image_path', type=str, help='Path to the input image')
parser.add_argument('output_path', type=str, help='Path to save the output image')
parser.add_argument('--params', nargs='+', type=int, help='Compression parameters')

# Parse the arguments
args = parser.parse_args()

# Read the input image
image = cv2.imread(args.image_path)

# Prepare the compression parameters (if provided)
compression_params = args.params if args.params else []

# Save the image with the specified compression parameters
cv2.imwrite(args.output_path, image, compression_params)
```

**Reply by mstenta** [reply]
Perhaps `imwrite_params` would be a better name for the argument, since it provides options beyond just compression.

https://docs.opencv.org/4.10.0/d8/d6a/group__imgcodecs__flags.html

It's also interesting because params needs to be a list of pairs like `[setting1_name, setting1_value, setting2_name, setting2_value]`.

So to disable TIFF compression, it looks like this (see my comment below where I tested this):

`cv.imwrite(output, panorama, [cv.IMWRITE_TIFF_COMPRESSION, cv.IMWRITE_TIFF_COMPRESSION_NONE])`

Which is equivalent to:

`cv.imwrite(output, panorama, [259, 1])`

**Reply by lukasalexanderweber** [reply]
Yeah the list would be supported with nargs='+'

**Comment by mstenta** [comment]
> By default `cv2.imread` will convert a 16 bit, three channel image in `a.tif` to 8 bit as shown in the question.

Based on this, I wonder if the 16-bit to 8-bit conversion is actually happening in `imread()`, and maybe if we fix that then `imwrite()` will output 16-bit automatically. The SO answer I linked to above suggests that's the case. They don't add any parameters to `imwrite()`, and they get a 16-bit output.

I tested it by changing line 113 of `images.py` from `img = cv.imread(img_name)` to `img = cv.imread(img_name, cv.IMREAD_UNCHANGED)` (based on https://learnopencv.com/read-an-image-in-opencv-python-cpp/), but now I'm getting this error:

```
Traceback (most recent call last):
  File "/usr/local/bin/stitch", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stitching/cli/stitch.py", line 312, in main
    panorama = stitcher.stitch(images, feature_masks)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stitching/stitcher.py", line 100, in stitch
    imgs, features, matches = self.subset(imgs, features, matches)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stitching/stitcher.py", line 146, in subset
    indices = self.subsetter.subset(self.images.names, features, matches)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stitching/subsetter.py", line 29, in subset
    indices = self.get_indices_to_keep(features, matches)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stitching/subsetter.py", line 68, in get_indices_to_keep
    raise StitchingError(
stitching.stitching_error.StitchingError: No match exceeds the given confidence threshold.
                                 Do your images have enough overlap and common
                                 features? If yes, you might want to lower the
                                 'confidence_threshold' or try another
                                 'detector'.
```

Perhaps 16-bit breaks the stitching logic in other places? :thinking:

**Reply by mstenta** [reply]
Tested with other `--detector` options:

**sift**:

```
cv2.error: OpenCV(4.10.0) /io/opencv/modules/features2d/src/sift.dispatch.cpp:512: error: (-5:Bad argument) image is empty or has incorrect depth (!=CV_8U) in function 'detectAndCompute'
```

**brisk**:

```
cv2.error: OpenCV(4.10.0) /io/opencv/modules/imgproc/src/color.simd_helpers.hpp:92: error: (-2:Unspecified error) in function 'cv::impl::{anonymous}::CvtHelper<VScn, VDcn, VDepth, sizePolicy>::CvtHelper(cv::InputArray, cv::OutputArray, int) [with VScn = cv::impl::{anonymous}::Set<3, 4>; VDcn = cv::impl::{anonymous}::Set<1>; VDepth = cv::impl::{anonymous}::Set<0, 2, 5>; cv::impl::{anonymous}::SizePolicy sizePolicy = cv::impl::<unnamed>::NONE; cv::InputArray = const cv::_InputArray&; cv::OutputArray = const cv::_OutputArray&]'
> Invalid number of channels in input image:
>     'VScn::contains(scn)'
> where
>     'scn' is 1
```

**akaze**:

```
cv2.error: OpenCV(4.10.0) /io/opencv/modules/stitching/src/exposure_compensate.cpp:564: error: (-215:Assertion failed) _image.type() == CV_8UC3 in function 'apply'
```

**Reply by mstenta** [reply]
I opened an issue for 16-bit support specifically: https://github.com/OpenStitching/stitching/issues/226

**Comment by mstenta** [comment]
Setting aside the 16-bit question... I tested disabling TIFF compression by changing line 313 of `stitch.py` from `cv.imwrite(output, panorama)` to `cv.imwrite(output, panorama, [cv.IMWRITE_TIFF_COMPRESSION, cv.IMWRITE_TIFF_COMPRESSION_NONE])`.

**Result: it worked!**

```
$ exiftool -compression test.tif
Compression                     : Uncompressed
```

And the output filesize is now 65.7 MB (instead of 26.4 MB with default LZW compression).

**Comment by mstenta** [comment]
@lukasalexanderweber I have created an `imwrite_params` branch on my fork, which adds an `--imwrite_params` CLI argument.

https://github.com/OpenStitching/stitching/compare/main...mstenta:stitching:imwrite_params

I tested disabling TIFF compression with this locally and it worked. The command looks like this:

`stitch ./*.tif --imwrite_params 259 1 --output test.tif`

I would be happy to open a pull request, but I have not run the tests, nor added any new tests for this. Would it help to open a "draft" PR as a first step at least? I defer to you on your preference as maintainer. :-)

**Reply by lukasalexanderweber** [reply]
Great! Just open a PR, it will run the tests automatically, we can add a new test afterwards

**Reply by mstenta** [reply]
Draft PR: https://github.com/OpenStitching/stitching/pull/225
